### PR TITLE
[WIP] Upgrade select2 to stable version 4.0.4 and fix compatibility issues in Civi

### PR DESCRIPTION
**Pull request by monishdeb**
Overview
----------------------------------------
This is patch is about using the latest stable select2/4.0.4 and fixing the  compatibility issues in Civi
  
Before
----------------------------------------
This is how single and multi-select worked in 3.5
test-multiple-before

After
----------------------------------------
test-multiple-after

Technical Details
----------------------------------------
This PR is the first step to make select2 accessible. Next step would be to migrate civicrm specific fixes in select2/3.5 if required. And then bring wooselect accessibility fixes in select2 library - https://github.com/select2/select2/compare/master...woocommerce:master

---

 * [CRM-21263: Upgrade select2 to stable version 4.0.4 and fix compatibility issues in Civi](https://issues.civicrm.org/jira/browse/CRM-21263)

### CODE REVIEW

**Comment by monishdeb** [comment]
ping @colemanw

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by monishdeb** [comment]
Encountered an issue with autocomplete widget. Marking WIP and will update the PR with fix.

File: js/Common.js [hunk 1]
```diff
@@ -542,7 +542,7 @@ if (!CRM.vars) CRM.vars = {};
             CRM.api3(entity, 'create', $.extend({name: e.object.term}, $el.data('api-params').params || {}))
               .done(function(created) {
                 var
-                  val = $el.select2('val'),
+                  val = $el.select2().val(),
```
Comment (colemanw) [inline js/Common.js]: I think this should simply be `val = $el.val(),`

**Review by colemanw — CHANGES_REQUESTED** [review]
I think this is a while off from addressing all the compatibility issues. The biggest one is that select2 v4 does not work on `<input>` elements at all, and will always post form data back to the server as an array. There are many spots in our forms that are expecting the old behavior of data posted as comma-separated strings.
Other issues that come to mind are entityRef fields, the "create new contact" box, and the `crm-action-menu` class.

**Comment by JoeMurray** [comment]
Thanks, @colemanw . Do you have a recommendation or preference for how we move this along while keeping the PR of a good size for review and QA the way @eileenmcnaughton likes? Should we do lots of commits against this one PR? Have a series of smaller PRs that pick away at changing one type of use of select2 to the new library at a time, renaming the widget type as part of this process? Other?

**Comment by jusfreeman** [comment]
@JoeMurray @monishdeb great job on this one, long awaited upgrade.

**Comment by eileenmcnaughton** [comment]
Re QA - my guess is that this will involve a susbstantial amount of time & you might be best to have a contract with @colemanw for that.

**Comment by JoeMurray** [comment]
Unfortunately we're doing this pro bono or I would definitely set up a contract. We have offers from some JAWS users to test.

**Comment by colemanw** [comment]
I am happy to work with Monish on this one. I feel a special kinship to select2 since I'm the one who originally pushed that library into core and painstakingly killed off all the old libraries it replaced.
Since this PR is not mergeable without fixing all the instances, I think we'll need to do them all in one PR. Let's just keep it rebased and in-sync with the latest master branch of Civi while we work on it. It's going to be a slog.

**Comment by monishdeb** [comment]
I agree with @colemanw as it won't be possible to chunk out small fixes into respective sub PRs, as the upgrade patch needs all fixes in one single PR, otherwise, it will cause issue in different places due to respective incompatible errors.

**Comment by eileenmcnaughton** [comment]
@colemanw @totten how does this change fit in our versioning? Is this a 6.0?

**Comment by colemanw** [comment]
@eileenmcnaughton this change will only affect extensions if they have javascript code that is tightly coupled to the internal workings of Select2, or if they rely on `$_POST` data from multiselect entityRefs. Should be very few if any extensions that do this. I think a message to the dev community will suffice.

**Comment by JoeMurray** [comment]
I've posted into extensions and dev channels on Mattermost.

**Comment by monishdeb** [comment]
Hi @colemanw, till now I have worked on these two issues- 
1. Partially fixed entityRef field, here's my [commit](https://github.com/civicrm/civicrm-core/pull/11969/commits/d838bed8713e742dcb943904542bc44e1d8b7468) Can you please help me in fixing the remaining issues:
  1.1 Placeholder doesn't show
 1.2 Create contact dialog box doesn't render below search field
 1.3 Escape text doesn't showup
 1.4 Email value don't get rendered below each Display name 
This is how it looks like after the partial fix:
test-multiple-after

2. Fixed the action-menu field to behave properly on selecting one or more record.
Screencast : 
test-multiple-after

**Comment by colemanw** [comment]
Yes some of that EntityRef stuff relies on JS code that's very tightly coupled with the internal workings of Select2 v3.5. So it will take some detangling to get it working with the new interface.

**Comment by eileenmcnaughton** [comment]
@monishdeb let's close this out until you have next worked on it - nothing lost by closing & re-opening (FYI if you re-open before doing a force push because then you can re-open the same PR , whereas if you re-open after you open a new PR - depending on the PR it's better to preserve comment history or bump to the top of the QA list)

**Pull request closed without merging**